Adding code for the data models.

marcboudreau opened the pull request:
This is an early contribution of code to model the various structs needed to build a graph of possible moves in an attempt to find the solution.

Commit:
Adding code for the data models.

danedmunds commented on board.go:
```diff
@@ -0,0 +1,141 @@
+// Direction constants.
+const (
+	Up Orientation = iota
+	Right
```
Do these get assigned somehow?

marcboudreau replied:
Yeah, it's one of those weird Golang things.  In the absence of an assignment, they are assigned the value of the above element + 1.

danedmunds replied:
cool!

danedmunds commented on board.go:
```diff
@@ -0,0 +1,141 @@
+
+// GetIndex calculates the index of a specific location within the pieces
+// slice.
+func (p *Board) GetIndex(x, y int) (int, error) {
```
Instead of doing this fancy thing, why not use a 2D slice?

marcboudreau replied:
I'm not really a fan of those, since it's not really a 2D slice, it's more of a slice of slices.   That way each subslice *can* be a different size.  Yuk!

marcboudreau replied:
Let me look into this a bit more before considering my reply...

danedmunds commented on board.go:
```diff
@@ -0,0 +1,141 @@
+			if index, err := p.GetIndex(x+i, y+j); err != nil {
+				return err
+			} else if p.slots[index] != nil {
+				return fmt.Errorf("Piece overlaps at %d, %d\n", x+i, y+j)
```
Maybe spit out the offending pieces?

marcboudreau replied:
Yeah that's a good point.  We should have some kind of unique identifier attached to the pieces.  That will be especially useful when we try to use the answer of this program to physically solve the puzzle.

danedmunds commented on board.go:
```diff
@@ -0,0 +1,141 @@
+			if index, err := p.GetIndex(x+i, y+j); err != nil {
+				return err
+			} else if p.slots[index] != nil && p.slots[index] != piece {
+				return errors.New("Cannot move piece")
```
Cannot move piece {blah} in direction {blah} it would overlap {blah} ?

marcboudreau replied:
Good point.

danedmunds reviewed (approved):
Cool!
I think we should use a 2D slice but do whatever! :)

marcboudreau commented:
Added some unit tests to cover the code I introduced.  Plus for kicks, I've hooked this repo up to TravisCI just to try something different.

https://travis-ci.org/marcboudreau/sunset-puzzle
